### ISSUE TITLE
Intersection over union not respecting class labels

### ISSUE THREAD

**Issue by Immeralcode** [Lightning-AI/torchmetrics#1921]
## 🐛 Bug

When using the IntersectionOverUnion function in torchmetrics.detection, the returned values do not respect the class labels. All the other boxes in the image are set to 0 when calculating the class IoU and it artificially deflates the IoU score. The function needs to either ignore boxes that do not have the same label from the class IoU score, or the function needs to drop all the 0's it inserts to the tensor before calculating the mean.

### To Reproduce

```python
import torch
from torchmetrics.detection import IntersectionOverUnion, MeanAveragePrecision

preds = [
   {
        "boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00], [298.55, 98.96, 314.97, 151.79]]),
        "scores": torch.tensor([0.236,0.56]),
        "labels": torch.tensor([4, 5]),
    }
]

target = [
    {
       "boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00], [298.55, 98.96, 314.97, 151.79]]),
       "labels": torch.tensor([4, 5]),
   }
]

IoU = IntersectionOverUnion(class_metrics=True, iou_threshold=.75, respect_labels=True)

demo_iou = IoU(preds, target)

print(demo_iou)

{'iou': tensor(1.), 'iou/cl_4': tensor(0.9307), 'iou/cl_5': tensor(0.9307)}
```

### Expected behavior

In the code above the code should return class based IoU's of 1, however it appears that instead of only using class 4 boxes for calculating IoU for class 4 it is including all boxes in the predictions for each class in the target. If you alter the first label in the target from 4 to 3 you get the following result:

{'iou': tensor(0.2500), 'iou/cl_3': tensor(0.), 'iou/cl_5': tensor(0.5000)}

since the target is still torch.tensor([4, 5]) we would expect an IoU for label 3 of 0, but the IoU for label 5 also drops to .5 when it's target and prediction boxes are exactly equal.

### Environment
torch                    2.0.1 (from pip)
torchmetrics             1.0.0 (from pip)
python 3.10 from conda

### Additional context

This was found during development for a production model which kept returning IoU scores of .1756 for a class which has a single great prediction box with true IoU of ~.88. The issue initially found was that all the other boxes, which were each labeled as a different class, were being included with scores of 0 in the mean calculation for the class IoU. Global IoU score also seems artificially reduced by the error.
